### ISSUE TITLE
Detached, closed docklet partially reappears when launching the IDE

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#3287]
Starting the IDE with any of the docklets (Documents, Post Window, Help Browser) being **detached and closed** will show up as a *ghost* window (empty and unclosable).

![deepinscreenshot_select-area_20180410201221](https://user-images.githubusercontent.com/271068/38586247-5869c160-3d1d-11e8-8c9b-8059c59a8402.png)

 
### Reproducer

- open, **detach** and **close** any docklet (View -> Docklets)
- close the IDE
- open the IDE again
- most of the time, see the *ghost* window in place of the closed docklet location

This happens at least on linux with some window manager (bug appears using `xfce4`, `icewm`, **not** with `awesome-wm`, ...).

------

Here, assuming the help browser is undocked --

It looks like the IDE's startup process always opens a window for the help browser. Then, if the IDE config file records that the help browser was closed when it last quit, it closes the browser window.

I found in the past, intermittently, that the IDE would try to close the help window before opening it -- leaving a non-functioning, non-painted window open. Before I can even close the window, I have to press Ctrl-D to load the main help page. Of course, the reason why I keep the help window closed is so that I don't have to index the help files unless I'm requesting help! So, this wastes time.

Since upgrading to 3.9 beta (built from source), I'm now finding that this race condition occurs roughly 90% of the time when I launch the IDE. So, it's changing from an occasional irritant into a constant WTF.

It's *reeeeeaaaalllllllllyyyyyyyy* annoying.

Ubuntu Studio 16.04, xfce 4.12, Qt 5.5.1 (from Ubuntu packages).

**Comment by telephon** [comment]
I can reproduce this on macos 10.11.6

**Comment by telephon** [comment]
goes away when I recompile the class library.

**Comment by jamshark70** [comment]
Still 80-90% of the time. It's really driving me nuts.

Where is the code to initialize the help browser? I may have a little more time coming up and... believe me... I am motivated to find this problem because IT'S RIDICULOUS.

**Comment by jamshark70** [comment]
I'm not sure of the right solution, but I'm looking around in the source code and thought I would record some notes here.

It looks like the relevant sequence is, more or less (all of these referring to IDE functions and classes):

1. `main()` creates an instance of `MainWindow`.

2. `MainWindow`'s constructor creates an instance of `HelpBrowserDocklet`.
   ```
       mHelpBrowserDocklet = new HelpBrowserDocklet(this);
       mHelpBrowserDocklet->setObjectName("help-dock");
       addDockWidget(Qt::RightDockWidgetArea, mHelpBrowserDocklet->dockWidget());
   ```

3. `main()` shows the main window.

4. `main()` gets geometry and dock state variables from a session, or last-saved defaults, and does either `sessions->openSession()` or `win->restoreWindowState()`.

5. AFAICS both of these paths eventually get around to `template <class T> void MainWindow::restoreWindowState( T * settings )`, which includes `restoreDetachedState( mHelpBrowserDocklet, detachedData );`.

6. `Docklet::restoreDetachedState()` sets the help window's visible state.

*If* these execute in sequence, then the help window would be created before restoring its state -- so the behavior I'm observing would be impossible. Therefore, at one or more stages, Qt actions are going off into different threads. But I don't see in our source code where this is happening -- so that must be in Qt internals, and I'm not sure we have a lot of control over it.

At this point, I'm out of my depth. I suppose we need to delay the part of `main()` that restores window geometry (https://github.com/supercollider/supercollider/blob/develop/editors/sc-ide/core/main.cpp#L103-L117), or somehow sync it to creating and showing the main window and all of its subcomponents.

Suggestions welcome.

**Comment by jamshark70** [comment]
I'm asking the experts. https://forum.qt.io/topic/86867/qdockwidget-visibility-sync-problem

**Comment by jamshark70** [comment]
https://forum.qt.io/post/436993:

> From a very quick look, I'd say that you should try delaying the settings loading to just after the event loop started. For that use a QTimer::singleShot (the overload that fits best) with 0 as delay.

http://doc.qt.io/qt-5/qtimer.html#singleShot -- with a few overloads, here's the first.

> void QTimer::singleShot(int msec, const QObject *receiver, const char *member)
> 
> This static function calls a slot after a given time interval.
> 
> It is very convenient to use this function because you do not need to bother with a timerEvent or create a local QTimer object.

So, basically, pop all of this out into a slot function and schedule it.

```
    QString startSessionName = settings->value("IDE/startWithSession").toString();
    if (startSessionName == "last") {
        QString lastSession = sessions->lastSession();
        if (!lastSession.isEmpty()) {
            sessions->openSession(lastSession);
        }
    }
    else if (!startSessionName.isEmpty()) {
        sessions->openSession(startSessionName);
    }

    if (!sessions->currentSession()) {
        win->restoreWindowState();
        sessions->newSession();
    }
```

There are probably 10 different ways to do this, and if I do it, it'll be wrong. So... suggestions?

**Comment by jamshark70** [comment]
Thinking out loud (and sorry if these are remedial C++ questions -- I *badly* want to fix this issue, but I don't understand how to get data from one part of the program to another -- I understand function calls but scheduling a slot on a timer is not a function call with arguments):

`settings` and `sessions` are available as Main member variables. So most of that block could be ported to a Main::initWindow function pretty easily (I guess).

`win->restoreWindowState();` -- `win`, inconveniently, is a local variable in `main()`. Ahh... second look... I could use `MainWindow::instance()` pretty much anywhere, right? Hmm...

Also, I have a feeling these bits should also be scheduled:

```
    foreach (QString argument, arguments) {
        main->documentManager()->open(argument);
    }

    win->restoreDocuments();
```

The document manager is a member variable (OK), but `arguments` is not. And there's no `Class::instance()` for arguments.

So I need to pass at least one `main()` local entity -- `arguments` -- into the slot function to handle the part of the init that should be delayed.

What is the best way, in C++, to do this?

Again, sorry if this is a dumb question, but I honestly don't know if this is a simple matter of moving some code into a function and jiggering a couple of variables, or if it was a mistake in the original code to initialize the window in main() rather than Main::something() (in which case it might need a more serious refactoring)... and it's a dodgy use of time to spend a few hours hacking up the first, only to be told in PR review that it was a totally wrong approach.

Thanks in advance.

**Comment by jamshark70** [comment]
Well. I might not be quite as much of an idiot as I thought. I went ahead with a POC, and it didn't crash, and it *did* change the behavior, but... not exactly right.

https://github.com/jamshark70/supercollider/commit/1cd6643aa2938d2eff5f0cf6d0adbf79689d8af6

This creates a `void Main::initWindow()` function and adds it under `public Q_SLOTS` (indentation is wrong, I'll fix it later). Then, `QTimer::singleShot(0, main, SLOT(initWindow()));` *does* actually run the thing. (The POC doesn't deal with the `arguments` problem.)

The problem is that the help window pops up in the "detached, open" state, where the state persisted into the IDE config file is "detached, closed."

One explanation that fits the facts is:

The old sequence may have been 1/ open main window; 1a/ create help docklet (but it's not displayed yet); 2/ apply settings loaded from the config file; 3/ go into the main loop; 4/ display windows.

But the new sequence is probably 1/ open main window; 2/ go into the main loop; 3/ display help docklet -- and I guess here, this changes the settings to say that help should be visible; 4/ apply settings (which are now different from the settings on disk).

I'll quit for today. Unclear what to do next. If `mHelpBrowserDocklet = new HelpBrowserDocklet(this)` is changing settings asynchronously, I don't yet see where that's happening.

For usage, it's a bit of an improvement: now, I only have to close the help window (where before, I needed to 1/ switch focus to the code window 2/ ctrl-D 3/ close the help window). But it's not *right* yet.

**Comment by jamshark70** [comment]
With a few debugging statements: in `Docklet::restoreDetachedState`, the visibility flag from the settings *is* 0.

So I guess what's happening is that now the help window init is happening at a time when it's legitimate for the help browser to request contents, and the completion of that request ignores window visibility (since, normal behavior if the help window is closed is ctrl-D --> request, indexing help files --> completed, show help window).

Not sure how to circumvent that logic.

**Comment by jamshark70** [comment]
Huh. With more testing, it seems the init behavior is still indeterministic. Sometimes the window gets painted (incompletely) right away; sometimes it's invisible for a few seconds... which could be explained by the initial observation (sometimes, it's "make window --> set visibility," sometimes the other way around). Then, after the app loop starts, the IDE requests the help front page and the window completes.

I was hoping for one of two things:

- Window creation is immediate; applying the visibility setting is deferred.

- Or, both are deferred and execute synchronously.

Anybody...?

**Comment by jamshark70** [comment]
Yes, that's the reproducer (make sure to click the dock menu at top left of the panel and choose "detach" -- I think the problem doesn't happen if the help window is simply dragged away to float off of the main window -- it has to be a separate window with its own title bar).

I'll check my qt version when I get back home.

At this point, I also can't rule out interaction with the OS window manager. Maybe it works fine with some desktops but not others. I'm using the version of XFCE 4 that ships with Ubuntu Studio.

**Comment by jamshark70** [comment]
And... "Help window pops up"... The problem is that the window gets deactivated while another thread is in the process of painting the window... So you get a garbage window frame but you can't click on it. (The close button doesn't work because, internally, the window is inactive.) It isn't a matter of the IDE properly and fully initializing a help window contrary to the saved settings. It tries to hide the window but it's only half successful.

It's also sporadic. If you tried only once, you might have just been lucky that it worked that time.

**Comment by jamshark70** [comment]
Qt 5.5.1, from Ubuntu packages.

On my old computer, I was using some version of Qt that I downloaded from the Qt website. I had some other SC/Qt-related problems, and the advice at that time was to stick with the distro's packages (I guess because the packages would be tested with this OS version more thoroughly).

I'll also bring in my question about getting access to the settings. I was testing an approach that might help (delaying `restoreDetachedState()`), but I can't be sure if it helps because the settings object behaves differently in two different places.

In `MainWindow::restoreWindowState( T * settings )`, the `settings` pointer is passed in as an argument.

```
template <class T>
void MainWindow::restoreWindowState( T * settings )
{
    qDebug("------------ restore window state ------------");

    printf("MainWindow::restoreWindowState settings address = %p\n", settings);
    
    settings->beginGroup("mainWindow");
    QVariant varGeom = settings->value("geometry");
    QVariant varState = settings->value("state");
    QVariant varDetached = settings->value("detached");
    settings->endGroup();

    QByteArray geom = QByteArray::fromBase64( varGeom.value<QByteArray>() );
    QByteArray state = QByteArray::fromBase64( varState.value<QByteArray>() );
    QVariantMap detachedData = varDetached.value<QVariantMap>();

    printf("detachedData address = %p\n", detachedData);

    ... blah blah
}
```

In the function I was testing (https://github.com/jamshark70/supercollider/commit/6e4085beb76b462cece896daf01c3a41f78f7d2b), I get the settings pointer from `Main::instance()->settings()`, then:

```
  void MainWindow::restoreDocksDetachedStates()
  {
    // printf("MainWindow::restoreDocksDetachedStates\n");
    
    Settings::Manager *settings = Main::instance()->settings();
    printf("MainWindow::restoreDocksDetachedStates settings address = %p\n", settings);

    settings->beginGroup("mainWindow");
    QVariant varGeom = settings->value("geometry");
    QVariant varState = settings->value("state");
    QVariant varDetached = settings->value("detached");
    settings->endGroup();

    QByteArray geom = QByteArray::fromBase64( varGeom.value<QByteArray>() );
    QByteArray state = QByteArray::fromBase64( varState.value<QByteArray>() );
    QVariantMap detachedData = varDetached.value<QVariantMap>();
    printf("detachedData address = %p\n", detachedData);

    ... blah blah
  }
```

The settings pointer has the same address in both functions, but `detachedData` has a different pointer between the two(++), even though (AFAICS) the operations are the same: `settings->value("detached")` and then `.value<QVariantMap>()`.

(++) Note, in the committed version, `%p` prints `detachedData` -- I also tested just now with `&detachedData` but it doesn't make a difference. I still get different values in the two places -- and in three trials, the difference is 0x60, seems not accidental.

Why are identical, and presumably deterministic, operations getting different results?

*I'm wondering*... would it be worth refactoring SettingsManager to serialize everything into a human-readable format? The fact that we're using base64 data for window geometry makes it much harder (and pointlessly harder) to troubleshoot any issues involving window geometry. I have zero visibility into the settings... no idea at all what's in there.

**Comment by gusano** [comment]
> At this point, I also can't rule out interaction with the OS window manager. Maybe it works fine with some desktops but not others. I'm using the version of XFCE 4 that ships with Ubuntu Studio.

Oh, while I'm using `awesome-wm` I can't reproduce it (never happened in the last year) but as soon as I switch to `xfce4` it happens always.. This is insanely annoying indeed. I'll do a bit more research. Thanks!

**Comment by gusano** [comment]
@jamshark70 After digging a bit more into this, I discovered the following:

- it seems the bug is `yaml` related
- 7f05dd4d8 seems to be the commit introducing the bug
- building from 97fe18126 (that is one commit **before**) shows no bug (once you have overwritten the scide settings file, you need to attach the Help Browser, quit, open scide, detach Help Browser, quit and open again)
- the offending commit was added alongside `Version-3.9.0-beta1` tag, it seems that some weird git tag magic happened there which I'm not able to understand but that's OT

My (wild) guess is that when *detaching* the Help Browser, what is saved in scide settings is a corrupted string and therefore when re-opening scide, boom, a dead window appears until you open the Help Browser again.

Alas I don't have the skills to dive in the `yaml` related `c++` code but it might help.

**Comment by jamshark70** [comment]
Hm, I was about to respond to a comment that seemed to identify a yaml re-factoring that might be what broke this, but I guess that comment is deleted now...? I got a github e-mail for the comment, but now that I come to the page, nothing.

At the very least, thanks to @gusano for confirming that the issue is as irritating as I say it is ;)  I suppose it might be documentation solution, then: "SC-IDE is currently incompatible with XFCE."

If it turns out to be yaml-related, that would add a little more weight to my suggestion a couple of days ago to stop using base64 to serialize window geometry. The lack of visibility into these settings is not helping with troubleshooting.

**Comment by gusano** [comment]
@jamshark70 yeah sorry, I deleted my comment after realizing I was completely wrong :)
I'm not sure if it would be a good idea to remove the base64 serialization.. Still looking into it.

**Comment by gusano** [comment]
@jamshark70 I think I found a fix. As the problem appears for any *detached* docklet (and not only Help Browser) I'm gonna update the issue title and description if you don't mind..
